tensorflow/core/kernels/dml_resource_scatter_ops.cc

```diff
@@ -0,0 +1,437 @@
+    const Tensor params_tensor =
+        init_helper->GetParamsTensor(ctx->GetOpKernelContext());
+
+    const TensorShape& in_out_shape = params_tensor.shape();
+    const TensorShape& indices_shape = ctx->GetInputTensorShape(1);
+    const TensorShape& updates_shape = ctx->GetInputTensorShape(2);
+
+    DmlTensorInfo in_out_tensor;
+    in_out_tensor.desc = DmlTensorDesc::Create(params_tensor.dtype(),
+                                               in_out_shape, in_out_shape);
+
+    DmlTensorInfo indices_tensor;
+    indices_tensor.desc = DmlTensorDesc::Create(ctx->GetInputDataType(1),
+                                                indices_shape, indices_shape);
+
+    DmlTensorInfo updates_tensor;
+    updates_tensor.desc = DmlTensorDesc::Create(ctx->GetInputDataType(2),
+                                                updates_shape, updates_shape);
+
+    DmlKernelTensors tensors;
+    tensors.inputs = {in_out_tensor, indices_tensor, updates_tensor};
+    tensors.outputs = {in_out_tensor};
+
+    if (params_tensor.dtype() != DT_RESOURCE) {
+      // The input ref and the output ref must refer to the same memory
+      tensors.output_refs_forwarding = {0};
+    }
+
+    auto inputs = GetDmlTensorDescs(tensors.inputs);
+    auto graph = dml::Graph(ctx->GetDmlDevice());
+    auto input = dml::InputTensor(graph, 0, inputs[0]);
+    auto indices = dml::InputTensor(graph, 1, inputs[1]);
+    auto updates = dml::InputTensor(graph, 2, inputs[2]);
+
+    // First, perform the scatter on an empty tensor
+    auto result = dml::ScatterND(input, indices, updates, in_out_shape.dims(),
+                                 indices_shape.dims());
+
+    Microsoft::WRL::ComPtr<IDMLCompiledOperator> compiled_op =
+        graph.Compile(DML_EXECUTION_FLAG_NONE, {result});
+
+    Initialize(ctx, std::move(tensors), compiled_op.Get());
+  }
+
+  StatusOr<DmlGpuEvent> Compute(DmlKernelContext* ctx) const override {
+    auto init_helper = ctx->GetInitializationHelper<InitHelper>();
+
+    const Tensor params_tensor =
+        init_helper->GetParamsTensor(ctx->GetOpKernelContext());
+
+    // Create input buffers
+    D3D12BufferRegion input_buffers[] = {
+        ctx->CreateBufferForTensor(params_tensor),
+        ctx->CreateBufferForTensor(ctx->GetInputTensor(1)),
+        ctx->CreateBufferForTensor(ctx->GetInputTensor(2)),
+    };
+
+    // Create input bindings
+    absl::optional<DML_BUFFER_BINDING> input_bindings[] = {
+        input_buffers[0].GetBufferBinding(),
+        input_buffers[1].GetBufferBinding(),
+        input_buffers[2].GetBufferBinding(),
+    };
+
+    DmlGpuEvent gpu_event;
+
+    absl::optional<DML_BUFFER_BINDING> output_bindings[] = {
+        input_bindings[0],
+    };
+
+    auto status_or_event =
+        DmlKernel::Compute(ctx, input_bindings, output_bindings);
+    if (!status_or_event.ok()) {
+      return status_or_event;
+    }
+
+    gpu_event = status_or_event.ValueOrDie();
+
+    init_helper->Unlock();

```

Comment:
This is super sketchy: there are some returns earlier in the function that would prevent the mutex from unlocking. One could argue that an error here would be fatal anyway, but I'd feel better if we had an RAII type that unlocks when destroyed (i.e. regardless of how compute returns).

It looks like this is an issue in the dml_scatter_update_ops.cc file as well (presumably adopted here), so it would be great to see both places updated. It does look like these kernels are never cached, which is good, so we shouldn't see multiple unlocks at least.